src/aggregate/aggregate_request.c

```diff
@@ -1072,6 +1075,32 @@ int AREQ_Compile(AREQ *req, RedisModuleString **argv, int argc, QueryError *stat
     goto error;
   }
 
+  // Define if we need a depleter in the pipeline
+  if (IsAggregate(req) && HasWithCount(req) && !IsCount(req)) {
+    PLN_ArrangeStep *arng = AGPLN_GetArrangeStep(AREQ_AGGPlan(req));
+    bool isLimited = (arng && arng->isLimited && arng->limit > 0);
+
+    if (req->protocol == 2)
+    {
+      if (!HasSortBy(req) || (HasSortBy(req) && isLimited)) {
```

Comment:
Can be the same condition as with protocol 3? `if (isLimited)`? And then no need to check for protocol?